### PR TITLE
Separate coordinating and primary bytes in stats

### PR DESCRIPTION
**Pull request by Tim-Brooks**
Currently we combine coordinating and primary bytes into a single bucket
for indexing pressure stats. This makes sense for rejection logic.
However, for metrics it would be useful to separate them.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java [hunk 1]
```diff
@@ -179,7 +179,15 @@ protected TransportReplicationAction(Settings settings, String actionName, Trans
...
     protected void doExecute(Task task, Request request, ActionListener<Response> listener) {
         assert request.shardId() != null : "request shardId must be set";
-        new ReroutePhase((ReplicationTask) task, request, listener).run();
+        runReroutePhase(task, request, listener, true);
+    }
+
+    private void runReroutePhase(Task task, Request request, ActionListener<Response> listener, boolean executedByClient) {
+        try {
+            new ReroutePhase((ReplicationTask) task, request, listener, executedByClient).run();
+        } catch (Exception e) {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:188]: It this necessary? `ReroutePhase` is an `AbstractRunnable` so should be fine

File: server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java [hunk 2]
```diff
@@ -1103,25 +1119,29 @@ public void markShardCopyAsStaleIfNeeded(ShardId shardId, String allocationId, l
         private final R request;
         // Indicates if this primary shard request originated by a reroute on this local node.
         private final boolean sentFromLocalReroute;
+        private final boolean localRerouteInitiatedByNodeClient;
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:1122]: add a comment here?

File: server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java [hunk 3]
```diff
@@ -658,12 +667,18 @@ private IndexShard getIndexShard(final ShardId shardId) {
...
         private final ClusterStateObserver observer;
         private final AtomicBoolean finished = new AtomicBoolean();
 
         ReroutePhase(ReplicationTask task, Request request, ActionListener<Response> listener) {
+            this(task, request, listener, false);
+        }
+
+        ReroutePhase(ReplicationTask task, Request request, ActionListener<Response> listener, boolean executedByClient) {
             this.request = request;
+            this.executedByClient = executedByClient;
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:681]: I prefer the `initiatedByNodeClient`  naming here with a comment explaining why  we have  that boolean

File: server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java [hunk 4]
```diff
@@ -1103,25 +1119,29 @@ public void markShardCopyAsStaleIfNeeded(ShardId shardId, String allocationId, l
         private final R request;
         // Indicates if this primary shard request originated by a reroute on this local node.
         private final boolean sentFromLocalReroute;
+        private final boolean localRerouteInitiatedByNodeClient;
 
         public ConcreteShardRequest(Writeable.Reader<R> requestReader, StreamInput in) throws IOException {
             targetAllocationID = in.readString();
             primaryTerm  = in.readVLong();
             sentFromLocalReroute = false;
+            localRerouteInitiatedByNodeClient = false;
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:1128]: can we also assert (similar  to sentFromLocalReroute) that localRerouteInitiatedByNodeClient is never false when being serialized in writeTo?

File: server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java [hunk 5]
```diff
@@ -80,17 +80,22 @@ protected TransportWriteAction(Settings settings, String actionName, TransportSe
...
+            // If this primary request was received from a local reroute initiated by the node client, we
+            // must mark a new primary operation local to the coordinating node.
+            if (localRerouteInitiatedByNodeClient) {
+                return indexingPressure.markPrimaryOperationLocalToCoordinatingNodeStarted(primaryOperationSize(request));
+            } else {
+                return () -> {};
+            }
         } else {
-            return indexingPressure.markIndexingOperationStarted(primaryOperationSize(request), forceExecution);
+            // If this primary request was received directly from the network, we must mark a new primary operation.
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java:97]: maybe mention that this is the case during primary delegation, after the primary relocation handoff.

**Review by ywelsch — COMMENTED** [review]
I've left some minor comments, looking good o.w.

File: server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java (hunk 1 above)
Comment (tbrooks8) [reply server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: Probably not? But we also do this with `new AsyncPrimaryAction(request, listener, (ReplicationTask) task).run();` in `handlePrimaryRequest`. So I did it for consistency? I think Henning was concerned that a `RuntimeException` would leave unreleased bytes.

Comment (tbrooks8) [reply server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: I did change it to just `RuntimeException`.

Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: ok

**Review by ywelsch — APPROVED** [review]
LGTM

**Comment by tbrooks8** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-windows

**Comment by tbrooks8** [comment]
@elasticmachine run elasticsearch-ci/1

**Comment by tbrooks8** [comment]
After discussing with @ywelsch - I am merging this PR. `packaging-sample-windows` seems to be an infra issue. And I had a green build on that this morning prior to some cosmetic API changes which are verified by dedicated tests.

**Pull request merged**